core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java

```diff
@@ -13,86 +13,186 @@
+            if (expression instanceof Constant) {
+                return Optional.of(literalEncoder.toExpression(((Constant) expression).getValue(), expression.getType()));
+            }
+            else if (expression instanceof FieldDereference) {
+                FieldDereference fieldDereference = (FieldDereference) expression;
+                Optional<Expression> baseOpt = process(fieldDereference.getTarget());
+                return baseOpt.map(base -> new SubscriptExpression(base, new LongLiteral(Long.toString(fieldDereference.getField() + 1))));
             }
+            else if (expression instanceof Variable) {
+                String name = ((Variable) expression).getName();
+                return Optional.of(variableMappings.map(map -> map.get(name).toSymbolReference())
+                                                   .orElse(new SymbolReference(name)));
+            }
+            else if (expression instanceof Call) {
+                return processCall((Call) expression);
+            }
+            return Optional.empty();
+        }
 
-            if (expression instanceof Constant) {
-                return literalEncoder.toExpression(((Constant) expression).getValue(), expression.getType());
+        protected Optional<Expression> processCall(Call call)
+        {
+            if (LIKE_FUNCTION_NAME.equals(call.getName())) {
+                return translateLike(call.getArguments().get(0), call.getArguments().get(1));
+            }
+
+            QualifiedName qualifiedName = getQualifiedName(call);
+
+            List<Expression> arguments = call.getArguments().stream()
+                                             .map(argument -> ConnectorExpressionTranslator.translate(session, argument, metadata, literalEncoder, Optional.empty()))
+                                             .collect(Collectors.toList());
+
+            if (REGEXP_LIKE_FUNCTION_NAME.equalsIgnoreCase(call.getName()) && arguments.size() == 2) {
+                Expression pattern = arguments.get(1);
+                if (pattern instanceof StringLiteral) {
+                    Slice slice = ((StringLiteral) pattern).getSlice();
+                    if (RegexLibrary.JONI.equals(SystemSessionProperties.getRegexLibrary(session))) {
+                        JoniRegexp joniRegexp = JoniRegexpCasts.castVarcharToJoniRegexp(slice);
+                        arguments.set(1, literalEncoder.toExpression(joniRegexp, JoniRegexpType.JONI_REGEXP));
+                    }
+                    else if (RegexLibrary.RE2J.equals(SystemSessionProperties.getRegexLibrary(session))) {
+                        int dfaStatesLimit = SystemSessionProperties.getRe2JDfaStatesLimit(session);
+                        int dfaRetries = SystemSessionProperties.getRe2JDfaRetries(session);
+                        Re2JRegexp re2JRegExp = new Re2JRegexp(dfaStatesLimit, dfaRetries, slice);
+                        Type type = new Re2JRegexpType(dfaStatesLimit, dfaRetries);
```

Comment:
Obtain the type object from `TypeManager` via `getType(Re2JRegexpType.RE2J_REGEXP_SIGNATURE)` instead of constructing an instance manually.

Reply:
Done, I had to pass `TypeManager` all the way to the translator

Reply:
I've tried to add `TypeManager` to `PlanOptimizers`'s `@Inject` constructor and to pass it from there. I had to change ~15 files including tests, but then I got a runtime error since `PlanOptimizers` is bound at `CoordinatorModule`, while `TypeManager` is not bound yet.
I couldn't find another way to obtain `TypeManager`, any suggestions?

Reply:
Thanks!
I just made a single change to your suggestion - replaced `if (!actualType.equals(expectedType)) {` with `if (!typeCoercion.isTypeOnlyCoercion(actualType, expectedType)) {` because, for example, at 
`testCrossJoinLargeBuildSideDynamicFiltering()` we get `$internal$dynamic_filter_function` with a `varchar(9)` argument  while its `expectedType = varchar`. Is this change ok by you?

In addition, are you aware of an equivalently generic way to implement `io.trino.sql.planner.ConnectorExpressionTranslator.SqlToConnectorExpressionTranslator#visitFunctionCall` which still refers to `JoniRegexp` and `Re2JRegexp` specifically?